apollo-backend/src/main/java/io/logz/apollo/services/SlaveService.java

```diff
@@ -0,0 +1,155 @@
+
+@Singleton
+public class SlaveService {
+    private static final Logger logger = LoggerFactory.getLogger(SlaveService.class);
+
+    private final String slaveId;
+    private final ScheduledExecutorService keepaliveExecutorService;
+    private final Boolean isSlave;
+    private final List<Integer> environmentIds;
+    private final ApolloConfiguration apolloConfiguration;
+    private final SlaveDao slaveDao;
+    private final EnvironmentDao environmentDao;
+
+    private boolean started = false;
+
+    @Inject
+    public SlaveService(ApolloConfiguration apolloConfiguration, SlaveDao slaveDao, EnvironmentDao environmentDao) {
+        this.slaveDao = requireNonNull(slaveDao);
+        this.environmentDao = requireNonNull(environmentDao);
+        this.apolloConfiguration = requireNonNull(apolloConfiguration);
+
+        isSlave = apolloConfiguration.getSlave().isSlave();
+        environmentIds = parseEnvironmentIds();
+
+        if (isSlave && isEmpty(environmentIds)) {
+            logger.error("Slave must be bundled with a valid list of environments! Bailing..");
+            throw new RuntimeException("Could not understand slaves params");
+        }
+
+        slaveId = UUID.randomUUID().toString();
+        keepaliveExecutorService = Executors.newSingleThreadScheduledExecutor(new BasicThreadFactory.Builder()
+                                            .namingPattern("slave-keepalive-pinger")
+                                            .build());
+    }
+
+    @PostConstruct
+    public void start() {
+        cleanupUnusedSlaves();
+
+        if (isSlave) {
+            claimSlave();
+            keepaliveExecutorService.scheduleWithFixedDelay(this::keepAlive,
+                    0, apolloConfiguration.getSlave().getKeepaliveIntervalSeconds(),
+                    TimeUnit.SECONDS);
+
+            started = true;
+        }
+    }
+
+    @PreDestroy
+    public void stop() {
+        started = false;
+        keepaliveExecutorService.shutdownNow();
+        slaveDao.removeAllSlavesById(slaveId);
+    }
+
+    public Boolean isRunningInSlaveMode() {
+        return isSlave;
+    }
+
+    public List<Integer> getEnvironmentIds() {
+        return environmentIds;
+    }
+
+    public List<Integer> getAllValidSlavesEnvironmentIds() {
+        return slaveDao.getAllSlaves().stream()
+                       .filter(slave -> slave.getSecondsSinceLastKeepalive()
+                               <= apolloConfiguration.getSlave().getKeepaliveIntervalSeconds() * 2)
+                       .map(Slave::getEnvironmentId)
+                       .collect(Collectors.toList());
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    @VisibleForTesting
+    public List<Integer> getScopedEnvironments() {
+        if (isSlave) {
+            return getEnvironmentIds();
+        } else { // I am the master, need all unattended environments
+            List<Integer> ownedEnvironments = getAllValidSlavesEnvironmentIds();
```

Comment:
Merge the 2 together

Reply:
Which two?

Reply:
2 methods. (this method and getAllValidSlavesEnvironmentIds)